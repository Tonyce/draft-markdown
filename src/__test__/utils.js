import { EditorState } from "draft-js";

export const applyMDtoInlineStyleChange = editorState =>
  EditorState.set(editorState, {
    lastChangeType: "md-to-inline-style",
  });
