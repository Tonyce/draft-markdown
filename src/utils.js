import { Modifier, EditorState } from "draft-js";

export const getCurrentLine = editorState => {
  const { anchorOffset } = editorState.getSelection();
  const selection = editorState.getSelection().merge({ anchorOffset });
  const key = editorState.getSelection().getStartKey();

  return editorState
    .getCurrentContent()
    .getBlockForKey(key)
    .getText()
    .slice(0, selection.getFocusOffset());
};

export function addText(editorState, bufferText) {
  const contentState = Modifier.insertText(
    editorState.getCurrentContent(),
    editorState.getSelection(),
    bufferText
  );
  return EditorState.push(editorState, contentState, "insert-characters");
}

export function replaceText(editorState, bufferText) {
  const contentState = Modifier.replaceText(
    editorState.getCurrentContent(),
    editorState.getSelection(),
    bufferText
  );
  return EditorState.push(editorState, contentState, "insert-characters");
}
