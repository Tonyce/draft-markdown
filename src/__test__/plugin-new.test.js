import Draft, { EditorState, SelectionState, convertToRaw } from "draft-js";
import createMarkdownPlugin from "../";
import { applyMDtoInlineStyleChange } from "./utils";

describe("markdown", () => {
  it("should convert asteriks to bold text", () => {
    const { handleBeforeInput } = createMarkdownPlugin();
    const setEditorState = jest.fn();
    const before = EditorState.moveSelectionToEnd(
      EditorState.createWithContent(
        Draft.convertFromRaw({
          entityMap: {},
          blocks: [
            {
              key: "item1",
              text: "Some *text",
              type: "unstyled",
              depth: 0,
              inlineStyleRanges: [],
              entityRanges: [],
              data: {},
            },
          ],
        })
      )
    );
    expect(handleBeforeInput("*", before, { setEditorState })).toEqual(
      "handled"
    );
    const raw = convertToRaw(
      setEditorState.mock.calls[0][0].getCurrentContent()
    );
    expect(raw).toMatchSnapshot();
  });

  it("should not do anything to existing inline styles when within them", () => {
    const { handleBeforeInput } = createMarkdownPlugin();
    const setEditorState = jest.fn();
    const boldInlineStyleRange = {
      length: 4,
      offset: 5,
      style: "BOLD",
    };
    const before = EditorState.forceSelection(
      EditorState.createWithContent(
        Draft.convertFromRaw({
          entityMap: {},
          blocks: [
            {
              key: "item1",
              text: "Some text",
              type: "unstyled",
              depth: 0,
              inlineStyleRanges: [boldInlineStyleRange],
              entityRanges: [],
              data: {},
            },
          ],
        })
      ),
      new SelectionState({
        anchorKey: "item1",
        anchorOffset: 6,
        focusKey: "item1",
        focusOffset: 6,
        isBackward: false,
        hasFocus: true,
      })
    );
    expect(handleBeforeInput("a", before, { setEditorState })).toEqual(
      "not-handled"
    );
  });

  it("should not unstick inline styles if they were not added by md-to-inline-style change", () => {
    const { handleBeforeInput } = createMarkdownPlugin();
    const boldInlineStyleRange = {
      length: 4,
      offset: 5,
      style: "BOLD",
    };
    const editorState = EditorState.moveSelectionToEnd(
      EditorState.createWithContent(
        Draft.convertFromRaw({
          entityMap: {},
          blocks: [
            {
              key: "item1",
              text: "Some text",
              type: "unstyled",
              depth: 0,
              inlineStyleRanges: [boldInlineStyleRange],
              entityRanges: [],
              data: {},
            },
          ],
        })
      )
    );
    expect(handleBeforeInput("a", editorState, {})).toEqual("not-handled");
  });

  it("should not have sticky inline styles", () => {
    const { handleBeforeInput } = createMarkdownPlugin();
    const setEditorState = jest.fn();
    const boldInlineStyleRange = {
      length: 4,
      offset: 5,
      style: "BOLD",
    };
    const editorState = applyMDtoInlineStyleChange(
      EditorState.moveSelectionToEnd(
        EditorState.createWithContent(
          Draft.convertFromRaw({
            entityMap: {},
            blocks: [
              {
                key: "item1",
                text: "Some text",
                type: "unstyled",
                depth: 0,
                inlineStyleRanges: [boldInlineStyleRange],
                entityRanges: [],
                data: {},
              },
            ],
          })
        )
      )
    );

    expect(handleBeforeInput("a", editorState, { setEditorState })).toEqual(
      "handled"
    );
    const raw = convertToRaw(
      setEditorState.mock.calls[0][0].getCurrentContent()
    );
    expect(raw.blocks[0].inlineStyleRanges[0]).toEqual(boldInlineStyleRange);
    expect(raw).toMatchSnapshot();
  });

  it("should not have sticky inline styles after the line ending with styles", () => {
    const { handleBeforeInput } = createMarkdownPlugin();
    const setEditorState = jest.fn();
    const boldInlineStyleRange = {
      length: 4,
      offset: 5,
      style: "BOLD",
    };
    const editorState = applyMDtoInlineStyleChange(
      EditorState.moveSelectionToEnd(
        EditorState.createWithContent(
          Draft.convertFromRaw({
            entityMap: {},
            blocks: [
              {
                key: "item1",
                text: "Some text",
                type: "unstyled",
                depth: 0,
                inlineStyleRanges: [boldInlineStyleRange],
                entityRanges: [],
                data: {},
              },
              {
                key: "item2",
                text: "",
                type: "unstyled",
                depth: 0,
                inlineStyleRanges: [],
                entityRanges: [],
                data: {},
              },
            ],
          })
        )
      )
    );

    expect(handleBeforeInput("a", editorState, { setEditorState })).toEqual(
      "handled"
    );
    const raw = convertToRaw(
      setEditorState.mock.calls[0][0].getCurrentContent()
    );
    expect(raw.blocks[0].inlineStyleRanges[0]).toEqual(boldInlineStyleRange);
    expect(raw).toMatchSnapshot();
  });
});
