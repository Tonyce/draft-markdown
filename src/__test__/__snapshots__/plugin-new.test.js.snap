// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown should convert asteriks to bold text 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 4,
          "offset": 5,
          "style": "BOLD",
        },
      ],
      "key": "item1",
      "text": "Some text",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`markdown should not have sticky inline styles 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 4,
          "offset": 5,
          "style": "BOLD",
        },
      ],
      "key": "item1",
      "text": "Some texta",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`markdown should not have sticky inline styles after the line ending with styles 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 4,
          "offset": 5,
          "style": "BOLD",
        },
      ],
      "key": "item1",
      "text": "Some text",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "item2",
      "text": "a",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`
